### PR TITLE
refactor fallback logic for qc voter & dkg broker

### PR DESCRIPTION
**Pull request by kc1116**
This PR refactors the fallback logic located in the QC Voter & DKG Broker for client connections into common default fall back strategy.

### CODE REVIEW

File: module/fallbackstrategy.go
Comment (jordanschalm) [inline module/fallbackstrategy.go:10]: We should include the client index being used here, otherwise we risk a race condition. The user of `FallbackStrategy` wants to tell `FallbackStrategy`, "I tried to talk to client X but it didn't work". If we don't pass in "X", which client we're referring to, we are requiring implicit synchronization between `FallbackStrategy`'s internal state and its user's state.

File: module/fallbackstrategy.go
Comment (jordanschalm) [inline module/fallbackstrategy.go:11]: A `Success(int)` method would be useful in general, and I think is required to fully reproduce the "fallback-after-2-consecutive-failures" logic we are trying to encapsulate in this component (otherwise we don't know when failures are consecutive).

File: module/util/util.go
Comment (jordanschalm) [inline module/util/util.go:10]: To close the loop with the original comment, I think this component should fully implement the fallback logic (eg. if we observe 2 consecutive `Failure()` calls without an intervening `Success()` call, then increment the client index. I think we should also:
* give it a more descriptive name, eg: `AfterConsecutiveFailuresStrategy`
* update the godoc to explain the functionality in more detail than "manage the active client index", for example "cycle the active client in a round-robin manner each time n consecutive failures are observed for the same client"

Comment (jordanschalm) [reply module/util/util.go:10]: It would also probably make sense to move this into a implementation package similar to many of the other interfaces defined in `module`, eg `module/fallback/after_consecutive_failures.go`

File: module/util/util.go
Comment (jordanschalm) [inline module/util/util.go:13]: I suggest we protect this with a mutex. The two use-cases we have now are single-threaded, but it's easy to imagine uses of this component which would concurrently access a client index and invoke `Failure` from two separate threads.

**Review by jordanschalm — CHANGES_REQUESTED** [review]
Thanks for the progress here. 

My concern is that this seems to be dividing the logic being refactored into two components, rather than cleanly encapsulating it fully in the new component (details in comments). This reduces the value of refactoring at all, in my opinion. I left some suggestions to more clearly draw the line between components here, and fully encapsulate the logic being refactored.

File: module/retrymiddleware/after_consecutive_failures.go
Comment (jordanschalm) [inline module/retrymiddleware/after_consecutive_failures.go:18]: This resets the client index to 0 each time we submit a new request. Suppose we have clients [a,b,c,d], and a and b are unavailable. Then we'll need to wait for failures on a and b each time before we can successfully send a request to c.

Comment (kc1116) [reply module/retrymiddleware/after_consecutive_failures.go:18]: Good call out, better to take into account **_last successful client_**

Comment (kc1116) [reply module/retrymiddleware/after_consecutive_failures.go:18]: Now that i'm thinking about it more, if we have to reset client index to 0 or we will be breaking the guarantee of client priority. We would also probably want those failure calls for metrics and logging.

Comment (jordanschalm) [reply module/retrymiddleware/after_consecutive_failures.go:19]: I think we should prioritize picking a client that is most likely to result in a successful request as quickly as possible, which in my mind would be that last client which was successful in a previous request.

Comment (kc1116) [reply module/retrymiddleware/after_consecutive_failures.go]: https://github.com/onflow/flow-go/pull/1556/commits/41cf3995759685b89b6b25c509fa5b02fefc0555

File: module/epochs/qc_voter.go
Comment (jordanschalm) [inline module/epochs/qc_voter.go:158]: Rather than the `bool` argument, which changes the behaviour quite a bit, could we separate this into two functions?
```
updateContractClient(clientIndex) (int, module.QCContractClient) // cycles the active client, returns the new client and index
getInitialContractClient() module.QCContractClient // retrieves the active client
```

File: module/epochs/qc_voter.go
Comment (jordanschalm) [inline module/epochs/qc_voter.go:164]: To me it would be simpler / less code to use `voter. lastSuccessfulClientIndex` directly instead of the intermediary variable. That way we can encapsulate mutex-protected reads/writes of that variable in the `updateContractClient` and `getInitialContractClient` methods too. 

Not a blocker though.

**Review by jordanschalm — COMMENTED** [review]
Final blocker: please make sure the shared `lastSuccessfulClientIndex` variable is concurrency-safe - I'm good to merge once that is done. A few non-blocker suggestions as well.

File: module/epochs/qc_voter.go
Comment (kc1116) [reply module/epochs/qc_voter.go:164]: It's a little more verbose but more explicit

File: module/epochs/qc_voter.go
Comment (kc1116) [reply module/epochs/qc_voter.go]: https://github.com/onflow/flow-go/pull/1556/commits/c7e2a246eb64337505deac805fd3fc71d6b3d9ef

File: module/retrymiddleware/after_consecutive_failures.go
Comment (kc1116) [reply module/retrymiddleware/after_consecutive_failures.go:17]: https://github.com/onflow/flow-go/pull/1556/commits/c7e2a246eb64337505deac805fd3fc71d6b3d9ef

File: module/dkg/broker.go
Comment (kc1116) [reply module/dkg/broker.go]: https://github.com/onflow/flow-go/pull/1556/commits/c7e2a246eb64337505deac805fd3fc71d6b3d9ef

File: module/dkg/broker.go
Comment (durkmurder) [inline module/dkg/broker.go:148]: do we need attempts only for logging?

Comment (kc1116) [reply module/dkg/broker.go:148]: yes @tarakby requested we log the attempts when we were debugging DKG broadcast issues some time after retries were introduced.

**Review by durkmurder — APPROVED** [review]
Overall looks good. But be aware that I am not familiar with this part of code :)

File: module/epochs/qc_voter.go
Comment (jordanschalm) [inline module/epochs/qc_voter.go:64]:
```suggestion
		mu:              sync.Mutex{},
```
We're only using the lock part of `Unit`, better to just use a lock directly

File: module/dkg/broker.go
Comment (jordanschalm) [inline module/dkg/broker.go:313]: ⚠️  Reads of the shared variable `lastSuccessfulClientIndex ` need to be synchronized as well. 

Currently the code will prevent concurrent _writes_ to `lastSuccessfulClientIndex`, but will allow concurrently one write and any number of reads. 

In general under the Go memory model, this results in undefined behaviour, from the [Go Memory Model](https://golang.org/ref/mem#tmp_10):
> Programs that modify data being simultaneously accessed by multiple goroutines must serialize such access.

File: module/dkg/broker.go
Comment (durkmurder) [inline module/dkg/broker.go:43]: It's a bit confusing why it has an embedded lock and then uses `unit` for locking as well.

File: module/dkg/broker.go
Comment (durkmurder) [reply module/dkg/broker.go:313]: +1 on this, strictly speaking this might not a problem. Since the write is protected by `Mutex` we will observe changes in other goroutines. I haven't found proof for that behavior in memory model, but go core devs use this [trick](https://github.com/golang/go/blob/master/src/runtime/chan.go#L138) and expect such behavior.

Just to be clear, I am not saying that we should do this, we should just place an explicit lock and be sure that it's safe and obvious, this off-topic is more for knowledge sharing of interesting fact.

File: module/dkg/broker.go
Comment (kc1116) [reply module/dkg/broker.go]: https://github.com/onflow/flow-go/pull/1556/commits/57c05d77a5b4485866c7036db4ad16e4ec2a9588

File: module/dkg/broker.go
Comment (kc1116) [reply module/dkg/broker.go:314]: https://github.com/onflow/flow-go/pull/1556/commits/6f9ab2af46584eb4c8fbea55778896dcbe0e2d9c

**Review by jordanschalm — APPROVED** [review]

**Comment by kc1116** [comment]
bors merge

**Pull request merged**
